pkg/agent/cluster.go

```diff
@@ -0,0 +1,330 @@
+	// GET /v2/clusters/{cluster_zero_id}
+	getClusterParams := &installer.V2GetClusterParams{ClusterID: *czero.clusterID}
+	result, err := czero.API.Rest.Client.Installer.V2GetCluster(czero.Ctx, getClusterParams)
+	if err != nil {
+		return nil, err
+	}
+	cluster := result.Payload
+	return cluster, nil
+}
+
+// IsBootstrapComplete Determine if the cluster that agent installer
+// is installing has completed the bootstrap process.
+func (czero *Cluster) IsBootstrapComplete() (bool, error) {
+
+	agentRestAPILive, agentRestAPIErr := czero.API.Rest.IsRestAPILive()
+	if agentRestAPIErr != nil {
+		logrus.Debug("node zero Agent API is not available")
+		logrus.Debug(agentRestAPIErr)
+	}
+
+	clusterKubeAPILive, clusterKubeAPIErr := czero.API.Kube.IsKubeAPILive()
+	if clusterKubeAPIErr != nil {
+		logrus.Debug("cluster Kube API is not available")
+		logrus.Debug(clusterKubeAPIErr)
+	}
+
+	if clusterKubeAPILive {
+		// First time we see the cluster Kube API
+		if !czero.installHistory.ClusterKubeAPISeen {
+			logrus.Info("cluster Kube API Initialized")
+			czero.installHistory.ClusterKubeAPISeen = true
+		}
+
+		configmap, _ := czero.API.Kube.IsBootstrapConfigMapComplete()
+		if configmap {
+			logrus.Info("bootstrap configMap status is complete")
+			return true, nil
+		}
+	}
+
+	if agentRestAPILive {
+		// First time we see the agent Rest API
+		if !czero.installHistory.RestAPISeen {
+			logrus.Info("node zero Agent API Initialized")
+			czero.installHistory.RestAPISeen = true
+		}
+		logrus.Trace("getting cluster info from Node Zero Agent API")
+		clusterState, _ := czero.Get()
+
+		// TODO(lranjbar)[AGENT-172]: Add CheckHostValidations
+		// if !czero.installHistory.AgentHostValidationsPassed {
+		// 	validations := CheckHostValidations()
+		// 	if validations {
+		// 		czero.installHistory.AgentHostValidationsPassed = true
+		// 	}
+		// }
+
+		czero.PrintInstallStatus(clusterState)
+		czero.installHistory.RestAPICurrentClusterStatus = *clusterState.Status
+
+		// Update Install History object when we see these states
+		czero.updateInstallHistoryClusterStatus(clusterState)
+
+		stopped, _ := czero.HasStoppedInstalling(*clusterState.Status)
+		if stopped {
+			logrus.Error("cluster has stopped installing")
+			errored, _ := czero.HasErrored(*clusterState.Status)
+			if errored {
+				return false, errors.New("cluster installation has stopped due to errors")
+			}
+			return false, errors.New("cluster has stopped installing and/or insallation was cancelled")
+		}
+
+	}
+
+	// both API's are not available
+	if !agentRestAPILive && !clusterKubeAPILive {
+		logrus.Debug("current API Status: Node Zero Agent API: down, Cluster Kube API: down")
+		if !czero.installHistory.RestAPISeen && !czero.installHistory.ClusterKubeAPISeen {
+			logrus.Debug("nero zero Agent API never initialized. Cluster API never initialized")
+			logrus.Warn("unable to detect installation. Cluster install has either not initalized or was not started")
+			return false, nil
```

Comment:
Just wondering if its intentional to not return an error here. and also if its not an error that a Warning should be logged.

Reply:
It was.. I'm not sure if this is an error or a warning. More likely info..